### PR TITLE
feature: script mode for model class

### PR DESCRIPTION
**Pull request by evakravi**
*Issue #, if available:*

*Description of changes:*
This PR adds script mode support to the `Model` class. 

This was basically done by adding the parameters `source_dir, code_location, entry_point, container_log_level, dependencies, git_config` to the `Model` class. 

*Testing done:*
The changes to the `Model` class do not break any existing unit tests. In addition, new unit tests were added to simulate the script mode use case for the `Model` class, and confirm that the calls to `sagemaker.deploy_endpoint()`, `sagemaker.create_model()`, `sagemaker.create_endpoint_configuration()` and `s3` are the same for the `Model` class and `FrameworkModel` class when both use script mode. In addition, this PR confirms that `git_config` can be used with the new `Model` class. Integration tests will be introduced in a subsequent PR. 

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/model.py [hunk 1]
```diff
@@ -237,10 +389,83 @@ def prepare_container_def(
...
+            else:
+                self._upload_code(deploy_key_prefix, repack=False)
+            deploy_env.update(self._script_mode_env_vars())
         return sagemaker.container_def(
-            self.image_uri, self.model_data, self.env, image_config=self.image_config
+            self.image_uri, self.model_data, deploy_env, image_config=self.image_config
         )
 
+    def _upload_code(self, key_prefix, repack=False):
+        """Placeholder Docstring"""
```
Comment (JGuinegagne) [inline src/sagemaker/model.py:409]: Would you be able to update the docstring?

**Review by JGuinegagne — APPROVED** [review]
non-blocking question: why moving `**/test_model.py` to `**/model/test_model.py` ?

File: src/sagemaker/model.py [hunk 2]
```diff
@@ -114,6 +130,124 @@ def __init__(
...
+                    >>> |----- README.md
+                    >>> |----- src
+                    >>>         |----- inference.py
+                    >>>         |----- test.py
+
+                    You can assign entry_point='inference.py', source_dir='src'.
+            code_location (str): Name of the S3 bucket where custom code is
+                uploaded (default: None). If not specified, default bucket
+                created by ``sagemaker.session.Session`` is used.
+            entry_point (str): Path (absolute or relative) to the Python source
```
Comment (mufaddal-rohawala) [inline src/sagemaker/model.py:155]: need to add `(default: None)`

**Review by mufaddal-rohawala — APPROVED** [review]

File: src/sagemaker/model.py [hunk 3]
```diff
@@ -242,10 +394,90 @@ def prepare_container_def(
...
+        deploy_env = copy.deepcopy(self.env)
+        if self.source_dir or self.dependencies or self.entry_point or self.git_config:
+            if self.key_prefix or self.git_config:
+                self._upload_code(deploy_key_prefix, repack=False)
+            elif self.source_dir and self.entry_point:
+                self._upload_code(deploy_key_prefix, repack=True)
+            else:
+                self._upload_code(deploy_key_prefix, repack=False)
+            deploy_env.update(self._script_mode_env_vars())
         return sagemaker.container_def(
```
Comment (shreyapandit) [inline src/sagemaker/model.py:409]: Can we concise this as:
```
if self.source_dir and self.entry_point:
	self._upload_code(deploy_key_prefix, repack=True)
else:
	self._upload_code(deploy_key_prefix, repack=False)
```

and with that you can even use ternary to condense it into a single line further

File: src/sagemaker/model.py [hunk 4]
```diff
@@ -242,10 +394,90 @@ def prepare_container_def(
...
+                s3_key_prefix=key_prefix,
+                script=self.entry_point,
+                directory=self.source_dir,
+                dependencies=self.dependencies,
+            )
+
+        if repack and self.model_data is not None and self.entry_point is not None:
+            if isinstance(self.model_data, sagemaker.workflow.properties.Properties):
+                # model is not yet there, defer repacking to later during pipeline execution
+                return
```
Comment (shreyapandit) [inline src/sagemaker/model.py:439]: This feels similar to elsewhere in the code, can we in future pull this into a common utility?

**Review by shreyapandit — APPROVED** [review]

File: src/sagemaker/model.py (hunk 4 above)
Comment (evakravi) [reply src/sagemaker/model.py:439]: We have this logic only once in the code, though there's similar packaging code used for the `Estimator` classes.

**Pull request merged**